### PR TITLE
Test+fix DeferredSuppliers in $brooklyn:object

### PR DESCRIPTION
**Pull request by aledsage**

### CODE REVIEW

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java [hunk 1]
```diff
@@ -428,28 +430,44 @@ public DslObject(
...
+                        }
+                        protected boolean requiresTask(Object value) {
+                            return (value instanceof TaskAdaptable || value instanceof TaskFactory);
+                        }
+                        protected Object resolveValue(Object value) {
+                            if (value instanceof DeferredSupplier) {
+                                return ((DeferredSupplier<?>) value).get();
                             }
+                            return value;
                         }
```
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: Good catch. Lots of repetition here, however. Maybe:
```
final Iterator<Object> taskValues = input.iterator();
Function<Object, Object> resolver = new Function<Object, Object>() {
    @Override public Object apply(Object value) {
        Object resolvedValue = requiresTask(value) ? taskValues.next() : resolveValue(value);
        return resolvedValue;
    }
};
Map<String, Object> resolvedConfig = Maps.transformValues(config, resolver);
Map<String, Object> resolvedFields = Maps.transformValues(fields, resolver);
List<Object> resolvedConstructorArgs = Lists.transform(constructorArgs, resolver);
List<Object> resolvedFactoryMethodArgs = Lists.transform(factoryMethodArgs, resolver);
```

File: camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java [hunk 2]
```diff
@@ -84,6 +94,17 @@ public void setManagementContext(ManagementContext managementContext) {
...
+    // TODO Copy of org.apache.brooklyn.rest.security.PasswordHasher; but this module does not
+    // have access to brooklyn-rest-resources (and would create a circular reference to do so)
+    public static class PasswordHasher {
+        public static String sha256(String salt, String password) {
+            if (salt == null) salt = "";
+            byte[] bytes = (salt + password).getBytes(Charsets.UTF_8);
+            HashCode hash = Hashing.sha256().hashBytes(bytes);
+            return hash.toString();
+        }
+    }
```
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java]: Perfect candiate for `brooklyn-utils-common` ;o)

Comment (alasdairhodge) [inline camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ObjectsYamlTest.java]: Nice use-case, btw!

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java (hunk 1 above)
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: Whoops, `fields` should come before `config`!

**Comment by alasdairhodge** [comment]
Love it, epsecially the thorough test coverage. Just nitpicks from me really; merge away.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java (hunk 1 above)
Comment (aledsage) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: yes, nice :-)

**Comment by aledsage** [comment]
Thanks @alasdairhodge - I'm incorporated those comments and also made a couple more tidy-ups (see commits that change it to use the `TypeResolver` utility, and that removes the strange `TypeCoercion` that I commented on previously).

Assuming you're happy with that, I'll merge once jenkins build completes.

@grkvlt I'd appreciate you taking a look as well please to let me know if I've missed something about why the code was written that way originally.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java [hunk 3]
```diff
@@ -47,6 +47,7 @@
 import org.apache.brooklyn.core.mgmt.internal.ManagementContextInternal;
 import org.apache.brooklyn.core.mgmt.persist.DeserializingClassRenamesProvider;
 import org.apache.brooklyn.core.sensor.DependentConfiguration;
+import org.apache.brooklyn.util.collections.MutableList;
```
Comment (alasdairhodge) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/BrooklynDslCommon.java]: Ha! No longer spurious!

**Comment by alasdairhodge** [comment]
Lovely jubbly.

**Comment by grkvlt** [comment]
@aledsage this looks OK to me, as well. going to merge

**Comment by grkvlt** [comment]
@aledsage oh, and it was written that way because that was the idiom in use at the time - `Tasks.resolveDeepValue` didn't exist I believe, but I would have used that as well if refactoring

**Pull request merged**